File: extractors/timetable-extractor/src/main/java/com/example/TimetableExtractor.java
/*
 * Bus Timetable PDF → CSV Extractor  (Java / Apache PDFBox 3.x)
 * ==============================================================
 * Maven dependency
 *   <dependency>
 *     <groupId>org.apache.pdfbox</groupId>
 *     <artifactId>pdfbox</artifactId>
 *     <version>3.0.1</version>
 *   </dependency>
 *
 * Gradle dependency
 *   implementation 'org.apache.pdfbox:pdfbox:3.0.1'
 *
 * Compile with standalone jar (no build tool)
 *   javac -cp pdfbox-app-3.0.1.jar TimetableExtractor.java
 *
 * Run
 *   java -cp ".:pdfbox-app-3.0.1.jar" TimetableExtractor input.pdf [output.csv]
 *   java -cp ".;pdfbox-app-3.0.1.jar" TimetableExtractor input.pdf [output.csv]  (Windows)
 */

import org.apache.pdfbox.Loader;
import org.apache.pdfbox.pdmodel.PDDocument;
import org.apache.pdfbox.text.PDFTextStripper;

import java.io.*;
import java.nio.charset.StandardCharsets;
import java.nio.file.Path;
import java.util.*;
import java.util.regex.*;

public class TimetableExtractor {

    /** Matches time tokens such as 7:40 or 10:00. */
    private static final Pattern TIME_RE = Pattern.compile("\\b(\\d{1,2}:\\d{2})\\b");

    public static void main(String[] args) throws IOException {
        if (args.length < 1) {
            System.err.println("Usage: TimetableExtractor <input.pdf> [output.csv]");
            System.exit(1);
        }

        String pdfPath = args[0];
        String csvPath = args.length > 1
                ? args[1]
                : Path.of(pdfPath).toString().replaceAll("\\.pdf$", ".csv");

        List<String[]> rows = extractRows(pdfPath);

        if (rows.isEmpty()) {
            System.err.println("No timetable data found.");
            System.exit(1);
        }

        writeCsv(rows, csvPath);
        System.out.printf("✓ %d stops written to '%s'%n", rows.size(), csvPath);
    }

    /**
     * Extract all timetable rows from the PDF.
     *
     * <p>Strategy: strip the full text with position-aware sorting, then for each
     * line match all time tokens. Everything before the first token is the stop name;
     * everything that matches is a departure time. Lines with no tokens are skipped
     * (headers, footers, legend lines, etc.).
     */
    private static List<String[]> extractRows(String pdfPath) throws IOException {
        List<String[]> rows = new ArrayList<>();

        try (PDDocument doc = Loader.loadPDF(new File(pdfPath))) {
            PDFTextStripper stripper = new PDFTextStripper();
            stripper.setSortByPosition(true);   // preserve reading order
            String text = stripper.getText(doc);

            for (String rawLine : text.split("\\r?\\n")) {
                String line = rawLine.strip();
                Matcher matcher = TIME_RE.matcher(line);

                List<String> times = new ArrayList<>();
                int firstStart = -1;

                while (matcher.find()) {
                    if (firstStart < 0) firstStart = matcher.start();
                    times.add(matcher.group());
                }

                if (times.isEmpty() || firstStart < 0) continue;

                String stopName = line.substring(0, firstStart).strip();
                if (stopName.isEmpty()) continue;

                String[] row = new String[1 + times.size()];
                row[0] = stopName;
                for (int i = 0; i < times.size(); i++) row[i + 1] = times.get(i);
                rows.add(row);
            }
        }

        return rows;
    }

    /** Write extracted rows to a UTF-8 CSV file with a generated header. */
    private static void writeCsv(List<String[]> rows, String csvPath) throws IOException {
        int maxServices = rows.stream().mapToInt(r -> r.length - 1).max().orElse(0);

        String[] header = new String[1 + maxServices];
        header[0] = "Paragem";
        for (int i = 1; i <= maxServices; i++) header[i] = "Serviço_" + i;

        try (PrintWriter pw = new PrintWriter(
                new OutputStreamWriter(new FileOutputStream(csvPath), StandardCharsets.UTF_8))) {
            pw.println(toCsvLine(header));
            for (String[] row : rows) pw.println(toCsvLine(row));
        }
    }

    /** Minimal RFC 4180 CSV serialisation for a single row. */
    private static String toCsvLine(String[] fields) {
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < fields.length; i++) {
            if (i > 0) sb.append(',');
            String f = fields[i] != null ? fields[i] : "";
            if (f.contains(",") || f.contains("\"") || f.contains("\n")) {
                sb.append('"').append(f.replace("\"", "\"\"")).append('"');
            } else {
                sb.append(f);
            }
        }
        return sb.toString();
    }
}

File: extractors/timetable_extractor.py
#!/usr/bin/env python3
"""
Bus Timetable PDF → CSV Extractor  (Python)
============================================
Dependency
    pip install pdfplumber

Usage
    python timetable_extractor.py <input.pdf> [output.csv]
"""

import csv
import re
import sys
from pathlib import Path

try:
    import pdfplumber
except ImportError:
    sys.exit("Missing dependency. Run:  pip install pdfplumber")

# Matches time tokens such as 7:40 or 10:00
_TIME_RE = re.compile(r"\b(\d{1,2}:\d{2})\b")


def parse_timetable(pdf_path: str) -> list[list[str]]:
    """
    Open a PDF and extract timetable rows.

    Each row is [stop_name, time_1, time_2, ...].
    Lines that contain no time tokens (headers, footers, legends) are skipped.
    """
    rows: list[list[str]] = []

    with pdfplumber.open(pdf_path) as pdf:
        for page in pdf.pages:
            text = page.extract_text(x_tolerance=3, y_tolerance=3)
            if not text:
                continue

            for raw_line in text.splitlines():
                line = raw_line.strip()
                times: list[str] = []
                first_start: int | None = None

                for m in _TIME_RE.finditer(line):
                    if first_start is None:
                        first_start = m.start()
                    times.append(m.group())

                if not times or first_start is None:
                    continue

                stop_name = line[:first_start].strip()
                if stop_name:
                    rows.append([stop_name] + times)

    return rows


def write_csv(rows: list[list[str]], csv_path: str) -> None:
    """Write extracted rows to a UTF-8 CSV file with a generated header row."""
    if not rows:
        sys.exit("No timetable data found in the PDF.")

    max_services = max(len(r) - 1 for r in rows)
    header = ["Paragem"] + [f"Serviço_{i}" for i in range(1, max_services + 1)]

    # utf-8-sig adds a BOM so Excel opens the file correctly
    with open(csv_path, "w", newline="", encoding="utf-8-sig") as fh:
        writer = csv.writer(fh)
        writer.writerow(header)
        writer.writerows(rows)

    print(f"✓ {len(rows)} stops written to '{csv_path}'")


if __name__ == "__main__":
    if len(sys.argv) < 2:
        sys.exit("Usage: python timetable_extractor.py <input.pdf> [output.csv]")

    pdf_in  = sys.argv[1]
    csv_out = sys.argv[2] if len(sys.argv) > 2 else str(Path(pdf_in).with_suffix(".csv"))
    write_csv(parse_timetable(pdf_in), csv_out)
